2001-06-07  Nathan Sidwell  <nathan@codesourcery.com>

	PR c++/2914
	* decl.c (pushtag): Don't push into a complete type's scope.

--- gcc/cp/decl.c.jj	2002-07-04 00:26:24.000000000 +0200
+++ gcc/cp/decl.c	2002-09-11 19:33:27.000000000 +0200
@@ -2870,7 +2870,13 @@ pushtag (name, type, globalize)
 
   b = current_binding_level;
   while (b->tag_transparent
-	 || (globalize && b->parm_flag == 2))
+	 || (b->parm_flag == 2
+	     && (globalize
+		 /* We may be defining a new type in the initializer
+		    of a static member variable. We allow this when
+		    not pedantic, and it is particularly useful for
+		    type punning via an anonymous union. */
+		 || COMPLETE_TYPE_P (b->this_class))))
     b = b->level_chain;
 
   b->tags = tree_cons (name, type, b->tags);
